### PR TITLE
Create Job API

### PR DESCRIPTION
**Pull request by vamsi-amazon**
### Description
* Datasource extraction from query will be handled in next PR.
* Documentation : https://github.com/vamsi-amazon/sql/blob/create-job-api/docs/user/interfaces/asyncqueryinterface.rst
* In the current PR, job APIs will be disabled if user doesn't configure `plugins.query.executionengine.spark.config` config. Once the customer updates the setting, user needs a domain restart to enable Job APIs.[Suggestions are welcome]
* Integration tests are not added, as we need emr serverless set up.

Sample Request Responses.

### Job Creation API

```
curl --location 'http://ec2-54-70-149-21.us-west-2.compute.amazonaws.com:9200/_plugins/_query/_jobs' \
--header 'Content-Type: application/json' \
--data '{
    "query" : "select * from default.http_logs limit 10"
}'
```

```
{
  "jobId": "00fd796ut1a7eg0q"
}
```
### Job Result API:

 ```
 curl --location --request GET 'http://ec2-54-70-149-21.us-west-2.compute.amazonaws.com:9200/_plugins/_query/_jobs/00fd796ut1a7eg0q/result' \
--header 'Content-Type: application/json' \
--data '{
    "query" : "select * from default.http_logs limit 1"
}'
 ```
 
 ```
 {"status":"RUNNING"}
 ```
 
### Check List
- [x] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: spark/src/main/java/org/opensearch/sql/spark/jobs/JobManagementService.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.sql.spark.jobs;
+
+public interface JobManagementService {
+
+  String createJob(String query);
+
+  String getJob(String jobId);
```
Comment (YANG-DB) [inline spark/src/main/java/org/opensearch/sql/spark/jobs/JobManagementService.java:12]: can the name be changes to getJobStatus ? and return more details ?

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+  private DataSourceService dataSourceService;
+
+  public String dispatch(String applicationId, String query) {
+    String datasourceName = getDataSourceName(query);
+    return emrServerlessClient.startJobRun(
+        applicationId, query, getExecutionRoleARN(datasourceName), getFlintHelper(datasourceName));
+  }
+
+  private String getDataSourceName(String query) {
```
Comment (YANG-DB) [inline spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java:34]: can this be a separate datasource for glue or more generalized a catalog based datasource ?

File: spark/src/main/java/org/opensearch/sql/spark/client/EmrServerlessClient.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package org.opensearch.sql.spark.client;
+
+import org.opensearch.sql.spark.helper.FlintHelper;
+
+public interface EmrServerlessClient {
```
Comment (YANG-DB) [inline spark/src/main/java/org/opensearch/sql/spark/client/EmrServerlessClient.java:5]: can the name be more generic ?

File: spark/src/main/java/org/opensearch/sql/spark/client/EmrServerlessClientImpl.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+import com.amazonaws.services.emrserverless.model.SparkSubmit;
+import com.amazonaws.services.emrserverless.model.StartJobRunRequest;
+import com.amazonaws.services.emrserverless.model.StartJobRunResult;
+import java.util.Set;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.sql.spark.helper.FlintHelper;
+import org.opensearch.sql.spark.response.SparkResponse;
+
+public class EmrServerlessClientImpl implements EmrServerlessClient {
```
Comment (YANG-DB) [inline spark/src/main/java/org/opensearch/sql/spark/client/EmrServerlessClientImpl.java:25]: this is the concrete impl for the server-less client

File: spark/src/main/java/org/opensearch/sql/spark/config/SparkExecutionEngineConfig.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+package org.opensearch.sql.spark.config;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.google.gson.Gson;
+import lombok.Data;
+
+@Data
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class SparkExecutionEngineConfig {
```
Comment (YANG-DB) [inline spark/src/main/java/org/opensearch/sql/spark/config/SparkExecutionEngineConfig.java:14]: does the name represents the actual execution config or only the client job dispatch ?

**Comment by penghuo** [comment]
> Datasource extraction from query will be handled in next PR.

What if we ask user provide datasource in query?

**Comment by YANG-DB** [comment]
@vamsi-amazon can u plz add a swagger file for a consolidated view and documentation for this API ?

File: spark/src/main/java/org/opensearch/sql/spark/jobs/JobExecutorServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+                () ->
+                    SparkExecutionEngineConfig.toSparkExecutionEngineConfig(
+                        sparkExecutionEngineConfigString));
+    String jobId =
+        sparkQueryDispatcher.dispatch(
+            sparkExecutionEngineConfig.getApplicationId(),
+            createJobRequest.getQuery(),
+            sparkExecutionEngineConfig.getExecutionRoleARN());
+    jobMetadataStorageService.storeJobMetadata(
+        new JobMetadata(jobId, sparkExecutionEngineConfig.getApplicationId()));
```
Comment (penghuo) [inline spark/src/main/java/org/opensearch/sql/spark/jobs/JobExecutorServiceImpl.java:68]: qq, if we can make sure there are at most 2 applicationId (old, latest), we don't need .ql.metadata system index, right?

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+      throw new IllegalArgumentException(
+          String.format(
+              "Bad URI in indexstore configuration of the : %s datasoure.", datasourceName));
+    }
+  }
+
+  public JSONObject getQueryResponse(String applicationId, String jobId) {
+    GetJobRunResult getJobRunResult = emrServerlessClient.getJobRunResult(applicationId, jobId);
+    JSONObject result = new JSONObject();
+    if (getJobRunResult.getJobRun().getState().equals(JobRunState.SUCCESS.toString())) {
```
Comment (penghuo) [inline spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java:59]: add todo. fetch from query_result index first. if no result, fetch from emr-s.

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+  // TODO: Analyze given query
+  // Extract datasourceName
+  // Apply Authorizaiton.
+  private String getDataSourceName() {
+    return "my_glue";
+  }
+
+  private String getDataSourceRoleARN(DataSourceMetadata dataSourceMetadata) {
+    return dataSourceMetadata.getProperties().get("glue.auth.role_arn");
```
Comment (penghuo) [inline spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java:74]: why glue.xxx.xxx?

File: spark/src/main/java/org/opensearch/sql/spark/client/EmrServerlessClient.java (hunk 3 above)
Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/client/EmrServerlessClient.java:12]: Will change it to spark client. Makes sense and one of the implementations is EMRServerlessClientImpl.

File: spark/src/main/java/org/opensearch/sql/spark/config/SparkExecutionEngineConfig.java (hunk 5 above)
Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/config/SparkExecutionEngineConfig.java:14]: This carries spark configuration set by user.

File: spark/src/main/java/org/opensearch/sql/spark/transport/TransportGetQueryResultRequestAction.java [hunk 9]
```diff
@@ -21,22 +30,46 @@ public class TransportGetQueryResultRequestAction
...
       GetJobQueryResultActionRequest request,
       ActionListener<GetJobQueryResultActionResponse> listener) {
-    String responseContent = "job result";
-    listener.onResponse(new GetJobQueryResultActionResponse(responseContent));
+    try {
+      String jobId = request.getJobId();
+      JobExecutionResponse jobExecutionResponse = jobExecutorService.getJobResults(jobId);
+      if (!jobExecutionResponse.getStatus().equals("SUCCESS")) {
+        JSONObject jsonObject = new JSONObject();
+        jsonObject.put("status", jobExecutionResponse.getStatus());
```
[the remark this answers is not in the document]
Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/transport/TransportGetQueryResultRequestAction.java:58]: was aware of this. Trying to fix it. Have created a task for this.

File: spark/src/main/java/org/opensearch/sql/spark/jobs/JobExecutorServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+  public CreateJobResponse createJob(CreateJobRequest createJobRequest) {
+    validateSparkExecutionEngineSettings();
+    String sparkExecutionEngineConfigString =
+        settings.getSettingValue(SPARK_EXECUTION_ENGINE_CONFIG);
+    SparkExecutionEngineConfig sparkExecutionEngineConfig =
+        AccessController.doPrivileged(
+            (PrivilegedAction<SparkExecutionEngineConfig>)
+                () ->
+                    SparkExecutionEngineConfig.toSparkExecutionEngineConfig(
+                        sparkExecutionEngineConfigString));
```
[the remark this answers is not in the document]
Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/jobs/JobExecutorServiceImpl.java:61]: SparkExecutionEngineConfig contains applicationId which can change and it is picked up runtime during query execution. Probably I need to split up the settings.

File: spark/src/main/java/org/opensearch/sql/spark/jobs/JobExecutorServiceImpl.java (hunk 6 above)
Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/jobs/JobExecutorServiceImpl.java:68]: I am kind of anticipating some more use cases with job metadata. Probably I will keep this for now. 
Once we stabilize apis and don't see any concerns we can remove and replace with some simple logic if we don't need.

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java (hunk 7 above)
Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java:59]: Hey what if the emr-s is still in the process of writing the result and customer queries. There is a chance, I can get half-baked results. But if we can live with that, I can switch.

File: spark/src/main/java/org/opensearch/sql/spark/jobs/OpensearchJobMetadataStorageService.java [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+   */
+  public OpensearchJobMetadataStorageService(Client client, ClusterService clusterService) {
+    this.client = client;
+    this.clusterService = clusterService;
+  }
+
+  @Override
+  public void storeJobMetadata(JobMetadata jobMetadata) {
+    if (!this.clusterService.state().routingTable().hasIndex(JOB_METADATA_INDEX)) {
+      createDataSourcesIndex();
```
[the remark this answers is not in the document]
Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/jobs/OpensearchJobMetadataStorageService.java]: At max one query execution can fail, but will work during retry.

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java (hunk 8 above)
Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java:74]: Added TODO. will address in next PR.

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java (hunk 7 above)
Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java:59]: Added TODO.

**Comment by vamsi-amazon** [comment]
> @vamsi-amazon can u plz add a swagger file for a consolidated view and documentation for this API ?

@YANG-DB  will address this in a separate task.

**Pull request merged**